barretenberg/cpp/src/barretenberg/benchmark/client_ivc_bench/client_ivc.bench.cpp

```diff
@@ -46,24 +46,59 @@ class ClientIVCBench : public benchmark::Fixture {
         GoblinMockCircuits::construct_mock_function_circuit(function_circuit);
         ivc.initialize(function_circuit);
 
-        // Accumulate kernel circuit (first kernel mocked as simple circuit since no folding proofs yet)
-        Builder kernel_circuit{ ivc.goblin.op_queue };
-        GoblinMockCircuits::construct_mock_function_circuit(kernel_circuit);
-        auto kernel_fold_proof = ivc.accumulate(kernel_circuit);
-
         auto NUM_CIRCUITS = static_cast<size_t>(state.range(0));
         NUM_CIRCUITS -= 1; // Subtract one to account for the "initialization" round above
-        for (size_t circuit_idx = 0; circuit_idx < NUM_CIRCUITS; ++circuit_idx) {
 
-            // Accumulate function circuit
-            Builder function_circuit{ ivc.goblin.op_queue };
-            GoblinMockCircuits::construct_mock_function_circuit(function_circuit);
-            auto function_fold_proof = ivc.accumulate(function_circuit);
+        ClientIVC::FoldProof function_fold_proof;
+        ClientIVC::FoldProof kernel_fold_proof;
+        for (size_t circuit_idx = 0; circuit_idx < NUM_CIRCUITS; ++circuit_idx) {
+            std::array<Builder, 2> kernel_and_function_builders;
+            parallel_for(2, [&](size_t workload_index) {
+                if (workload_index == 0) {
+                    // Initialize with current op_queue
+                    kernel_and_function_builders[workload_index] = Builder(ivc.goblin.op_queue);
+                    // Compute kernel circuit
+                    if (circuit_idx == 0) {
+                        // First kernel is not really a kernel
+                        GoblinMockCircuits::construct_mock_function_circuit(
+                            kernel_and_function_builders[workload_index]);
+                    } else {
+                        // Actually kernel for next
+                        GoblinMockCircuits::construct_mock_folding_kernel(
+                            kernel_and_function_builders[workload_index], function_fold_proof, kernel_fold_proof);
+                    }
+                } else {
+                    // Construct function circuit in parallel
+                    // Initialize without op_queue
+                    kernel_and_function_builders[workload_index] = Builder();
+                    GoblinMockCircuits::construct_mock_function_circuit(kernel_and_function_builders[workload_index]);
+                }
+            });
+            // Accumulate kernel
+            kernel_fold_proof = ivc.accumulate(kernel_and_function_builders[0]);
+
+            // Prepend op_queue to function circuit
+            kernel_and_function_builders[1].op_queue->prepend_previous_queue(*ivc.goblin.op_queue);
```

Comment:
Because there was no functionality allowing such updates to the queue and it is needed for correctness (we need to prepend previous queue so that the one being accumulated is correct, but we don't care about the previous queue during function circuit construction)